R/moduleInstall.R

```diff
@@ -1,12 +1,338 @@
+
+  renv::consent() #Consent to doing stuff in certain paths: https://rstudio.github.io/renv/reference/consent.html We've changed at least the root and the cache so hopefully that means it won't be changing stuff in the locations it mentions
+  options(renv.verbose=TRUE) # More feedback wanted although this seems to do little
+  options(renv.config.install.verbose=TRUE)
+
+  if (!dir.exists(moduleLibrary))
+    if (!dir.create(moduleLibrary))
+      stop("failed to create moduleLibrary!")
+
+  # ensure this starts with a ., otherwise it's picked up by renv "helpers"
+  moduleLibraryTemp <- file.path(moduleLibrary, ".renv_temp")
+
+  if (!dir.exists(file.path(moduleLibraryTemp, "renv")))
+    dir.create(   file.path(moduleLibraryTemp, "renv"), recursive = TRUE)
+
+  lockFileModule <- file.path(modulePkg,         "renv.lock")
+  lockFileTemp   <- file.path(moduleLibraryTemp, "renv.lock")
+  file.copy(from = lockFileModule, to = lockFileTemp, overwrite = TRUE)
+
+  setupRenv(moduleLibrary)
+
+  # TODO: unclear whether this is necessary within JASP, or just within Rstudio.
+  # renv must be unaware of any other libPaths than the cache and the directory designated for the module.
+  # if there are other libPaths then renv may reuse pkgs from those other libPaths
+  # it does copy those pkg to the cache before symlinking them
+  # inspired by https://stackoverflow.com/a/36873741/4917834
+  # it does appear to be necessary within rstudio and when the pkgs from jasp-required-files are present in a libPath
+  old.lib.loc <- .libPaths()
+  on.exit(assign(".lib.loc", old.lib.loc,   envir = environment(.libPaths)))
+          assign(".lib.loc", moduleLibrary, envir = environment(.libPaths))
+
+
+  lib <- renv::paths[["library"]](project = moduleLibrary)
+  if (!dir.exists(lib))
+    dir.create(lib, recursive = TRUE)
+
+  renv::restore(project  = moduleLibraryTemp,
+                library  = moduleLibrary,
+                lockfile = lockFileTemp, clean = TRUE,
+                prompt   = prompt)
+
+  moduleInfo         <- getModuleInfo(modulePkg)
+  correctlyInstalled <- installModulePkg(modulePkg, moduleLibrary, prompt, moduleInfo)
+
+  if (correctlyInstalled)
+    writeMd5Sums(modulePkg, moduleLibrary)
+
+  renv::snapshot(
+    project  = moduleLibraryTemp,
+    lockfile = lockFileTemp,
+    packages = moduleInfo[["Package"]],
+    prompt   = prompt,
+    force    = TRUE # force is "safe" here because we only update the new module
+  )
+
+  # some checks here to assert that everything got installed correctly
+  if (!libraryMatchesLockfile(moduleLibraryTemp)) {
+
+    # do it again!
+    renv::restore(
+      project  = moduleLibraryTemp,
+      library  = moduleLibrary,
+      lockfile = lockFileTemp,
+      clean    = TRUE,
+      prompt   = prompt
+    )
+
+    if (!libraryMatchesLockfile(moduleLibraryTemp))
+      warning("Failed to recreate lock file of module!")
```

Comment:
It means that renv had to improvise and could not exactly replicate a lockfile as is. This can happen if you use package versions that are not available anywhere, or cannot install old packages from source without any available binary. In that scenario, renv will just try to make things work with other package versions (which I think is reasonable).

I don't think this will happen ever if:
- a module doesn't use a lockfile but only a DESCRIPTION
- a module updates its dependencies somewhat regularly (say twice a year).